sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BeamFnDataReadRunnerTest.java

```diff
@@ -122,334 +129,579 @@
+    public List<String> processedElements;
+
+    @Parameterized.Parameter(3)
+    public double fractionOfRemainder;
+
+    @Parameterized.Parameter(4)
+    public double elementProgress;
+
+    @Test
+    public void testElementSplit() throws Exception {
+      SplittingReceiver splittingReceiver = mock(SplittingReceiver.class);
+      BeamFnDataClient mockBeamFnDataClient = mock(BeamFnDataClient.class);
+      when(splittingReceiver.getProgress()).thenReturn(elementProgress);
+      when(splittingReceiver.trySplit(anyDouble())).thenReturn(mockSplitResult);
+      BeamFnDataReadRunner<String> readRunner =
+          createReadRunner(splittingReceiver, PTRANSFORM_ID, mockBeamFnDataClient);
+
+      assertEquals(
+          expectedResponse,
+          executeSplit(
+              readRunner,
+              PTRANSFORM_ID,
+              inputElements,
+              processedElements,
+              fractionOfRemainder,
+              Collections.EMPTY_LIST));
+    }
   }
 
-  @Test
-  public void testSplittingDownstreamReceiver() throws Exception {
-    SplitResult splitResult =
-        SplitResult.of(
-            BundleApplication.newBuilder().setInputId("primary").build(),
-            DelayedBundleApplication.newBuilder()
-                .setApplication(BundleApplication.newBuilder().setInputId("residual").build())
-                .build());
-    SplittingReceiver splittingReceiver = mock(SplittingReceiver.class);
-    when(splittingReceiver.getProgress()).thenReturn(0.3);
-    when(splittingReceiver.trySplit(anyDouble())).thenReturn(splitResult);
-    BeamFnDataReadRunner<String> readRunner = createReadRunner(splittingReceiver);
-
-    ProcessBundleSplitRequest request =
-        ProcessBundleSplitRequest.newBuilder()
-            .putDesiredSplits(
-                "pTransformId",
-                DesiredSplit.newBuilder()
-                    .setEstimatedInputElements(10)
-                    .setFractionOfRemainder(0.05)
-                    .build())
-            .build();
-    ProcessBundleSplitResponse.Builder responseBuilder = ProcessBundleSplitResponse.newBuilder();
-
-    // We will be "processing" the 'C' element, aka 2nd index
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("A"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("B"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("C"));
-    readRunner.trySplit(request, responseBuilder);
+  // Test different cases of element split with non-empty allowed split points.
+  @RunWith(Parameterized.class)
+  public static class ElementSplitWithAllowedSplitPointsTest {
+    @Parameterized.Parameters
+    public static Iterable<Object[]> data() {
+      return ImmutableList.<Object[]>builder()
+          // This is where we would like to split, when all split points are available.
+          .add(
+              new Object[] {
+                elementSplitResult(2L),
+                5L,
+                ImmutableList.of("A", "B", "C"),
```

Comment:
nit: If you want to better mirror the python version of the test, you could generate the list of elements in testElementSplittingWithAllowedSplitPoints based upon the "index"  you pass in instead of explicitly listing them (here and in the other parameterized versions)

nit: ordering the parameters the same way also makes for easier maintenance in the future (here and in the other parameterized versions)